### PR TITLE
Bug 1961233: Add DNS availability upgrade test

### PR DESCRIPTION
**Pull request by suleymanakbas91**
Add a new test to the upgrade test suite to validate DNS availability during and after an upgrade. The new test implements the simple [UpgradeTest interface](https://github.com/openshift/origin/blob/master/vendor/k8s.io/kubernetes/test/e2e/upgrades/upgrade.go#L46). 
- `test/e2e/upgrade/dns/dns.go`: 
(`Name()`) and (`DisplayName()`) provides a name for the test to be used by the framework (e.g. for namespace creation, printing in the logs). 
(`Setup()`) performs the intial setup before the upgrade starts. It retrieves DNS Service Cluster IP and creates a DaemonSet (`createDNSTestDaemonSet()`) which performs a simple dig command (`while true; do dig +short @CLUSTER_IP google.com || echo $(date) fail && sleep 1; done`) to check DNS availability using the obtained Cluster IP.
(`Test()`) is called during upgrade and the `done` channel is closed when the upgrade is finished. `validateDNSResults()` is the function called within `Test()` to test the availability during and after upgrade (i.e. after `done` is closed). It retrieves logs from DaemonSet Pods, counts number of sucessful and failed DNS queries, and calculates the success rate. It fails the test if the success rate is less than 99%. 
(`Teardown()`) is an empty function created only to implement the interface. Framework deletes the namespace created for this test after the test suit finishes. That's why there is no any other clean up need for this test.
- `test/e2e/upgrade/upgrade.go`: Add the new DNS upgrade test to the list of `AllTests` so that it gets executed as a part of ClusterUpgrade tests.

### CODE REVIEW

File: test/e2e/upgrade/dns/dns.go
Comment (brandisher) [inline test/e2e/upgrade/dns/dns.go:71]: It might be a good idea to add a `sleep` in here. Maybe `sleep 1` just to space out the lookups a bit. We should also consider whether caching is an important aspect here...for example, if the default caching is 3 seconds we could get a cached record for _n_ lookups before receiving a failure.

I also noticed that `dig` doesn't differentiate between a successful lookup and an `NXDOMAIN`.
```
$ dig +noall asdlkfhsdlfhsdaklhfklsjflklskhflsdflkadflhsdafkljhsdakjhfsdf.com
$ echo $?
0
$ dig +noall google.com
$ echo $?
0
```
The only problem with that is we have no way of programmitcally differentiating between a failed lookup for a domain that we know should be successful (e.g. google.com) and a domain that we know should not be successful. To solve this, it might make sense to just do:
```
while true; do dig +short @%s google.com || echo fail; done
```
Then when looping through the pod logs you can validate that a line is an IP address. If a line is "fail" or not an IP, you know something else went wrong and can fail accordingly. Hopefully that all makes sense!

**Review by brandisher — CHANGES_REQUESTED** [review]

File: test/e2e/upgrade/dns/dns.go
Comment (suleymanakbas91) [reply test/e2e/upgrade/dns/dns.go]: I appended a random UUID to the end so that it will be unique.

File: test/e2e/upgrade/upgrade.go
Comment (suleymanakbas91) [reply test/e2e/upgrade/upgrade.go]: Done.

File: test/e2e/upgrade/dns/dns.go
Comment (suleymanakbas91) [reply test/e2e/upgrade/dns/dns.go:108]: Nice tooling, but I'm not sure if we can use it here. We might use it to look for the `fail` keyword, but in our case, we should also check if there is anything else in the logs rather than IP addresses or `fail` keyword. Not sure how to achieve this with this tooling.

File: test/e2e/upgrade/dns/dns.go
Comment (suleymanakbas91) [reply test/e2e/upgrade/dns/dns.go]: I've applied the suggestion. It's much simpler and more efficient. Thanks!

File: test/e2e/upgrade/dns/dns.go
Comment (suleymanakbas91) [reply test/e2e/upgrade/dns/dns.go]: I've applied the suggestion. Thanks!

**Review by brandisher — APPROVED** [review]
This looks good to me! Ready for review by the broader team.

File: test/e2e/upgrade/dns/dns.go
Comment (candita) [inline test/e2e/upgrade/dns/dns.go:7]: We follow a conventional pkg ordering pattern of:
* single-name pkg alphabetically ordered
* blank line
* github pkgs alphabetically ordered
* blank line
* k8s pkgs alphabetically ordered
* blank line
* other pkgs in alphabetical order

File: test/e2e/upgrade/dns/dns.go
Comment (candita) [inline test/e2e/upgrade/dns/dns.go:45]: We want to test during an upgrade as well, you need to add a test for during the upgrade.

File: test/e2e/upgrade/dns/dns.go
Comment (candita) [inline test/e2e/upgrade/dns/dns.go:64]: Any reason you can't do this with just a pod?  https://github.com/openshift/origin/blob/master/test/extended/dns/dns.go#L30

File: test/e2e/upgrade/dns/dns.go
Comment (candita) [inline test/e2e/upgrade/dns/dns.go:67]: Anything you create using f.ClientSet, you will need to clean up later by deleting it, usually done in a `defer` block.

File: test/e2e/upgrade/dns/dns.go
Comment (candita) [inline test/e2e/upgrade/dns/dns.go:128]: I don't see where you created or deleted a namespace for the new objects?

File: test/e2e/upgrade/dns/dns.go
Comment (candita) [inline test/e2e/upgrade/dns/dns.go:92]: Overall, I think you should model this test re-using more from https://github.com/openshift/origin/blob/master/test/extended/dns/dns.go.

File: test/e2e/upgrade/dns/dns.go
Comment (suleymanakbas91) [reply test/e2e/upgrade/dns/dns.go:11]: Thanks! I re-ordered them based on this. Is there a place where I can see all the team conventions?

File: test/e2e/upgrade/dns/dns.go
Comment (suleymanakbas91) [reply test/e2e/upgrade/dns/dns.go:50]: Fixed. Now I run the test during and after upgrade.

File: test/e2e/upgrade/dns/dns.go
Comment (suleymanakbas91) [reply test/e2e/upgrade/dns/dns.go:132]: If the `SkipNamespaceCreation` is not set explicitly, a namespace is being created and deleted by the framework: https://github.com/openshift/origin/blob/master/vendor/k8s.io/kubernetes/test/e2e/framework/framework.go#L238

File: test/e2e/upgrade/dns/dns.go
Comment (suleymanakbas91) [reply test/e2e/upgrade/dns/dns.go:71]: The namespace being used here will already get deleted by the framework at the end of the test. That's why, AFAIU, deleting these test resources explicitly should not be required.

File: test/e2e/upgrade/dns/dns.go
Comment (suleymanakbas91) [reply test/e2e/upgrade/dns/dns.go:68]: No, but this was Miciah's idea to use a DaemonSet to test the DNS availability from all the Pods.

File: test/e2e/upgrade/dns/dns.go
Comment (candita) [reply test/e2e/upgrade/dns/dns.go:11]: Mostly just what `go fmt` and `goimports` do.  There are also https://github.com/openshift/enhancements/blob/master/CONVENTIONS.md#api and https://go.dev/doc/effective_go#formatting

File: test/e2e/upgrade/dns/dns.go
Comment (candita) [reply test/e2e/upgrade/dns/dns.go:96]: After discussion, I will leave this up to the approver/s.

File: test/e2e/upgrade/dns/dns.go
Comment (candita) [inline test/e2e/upgrade/dns/dns.go:69]: After discussion, it would be better if you could also echo a timestamp here when it fails.

**Comment by candita** [comment]
After discussion, @suleymanakbas91  if you could please check: https://github.com/openshift/origin/pull/26795/commits/c10e64e1f5f591c45ad15993dca4147b20b21a7d#r817073378

Otherwise looks good to me and you should contact someone from the OWNERS file in the #forum-api-review slack channel for approval.  This will be invaluable in our CI and upgrade improvement strategy.

/lgtm

File: test/e2e/upgrade/dns/dns.go
Comment (suleymanakbas91) [reply test/e2e/upgrade/dns/dns.go]: Now, I'm also printing a timestamp when it fails.

File: test/e2e/upgrade/dns/dns.go
Comment (deads2k) [inline test/e2e/upgrade/dns/dns.go:121]: how will you know node had the problem?

**Comment by deads2k** [comment]
/approve

this is a great test to add.  I've got a comment about improving the error message

/hold in case you want to address the comment.


I also suggest adding this as an invariant test for *all* openshift-tests since DNS should work during all the tests we run.

File: test/e2e/upgrade/dns/dns.go
Comment (suleymanakbas91) [reply test/e2e/upgrade/dns/dns.go]: Now I'm printing the node name as well.

**Comment by brandisher** [comment]
Retesting as the failures don't appear to be related to this change.

/retest

**Pull request merged**
